src/Orleans.Core/Providers/IGrainStorage.cs

```diff
@@ -0,0 +1,146 @@
+{
+    /// <summary>
+    /// Interface to be implemented for a storage able to read and write Orleans grain state data.
+    /// </summary>
+    public interface IGrainStorage
+    {
+        /// <summary>Read data function for this storage instance.</summary>
+        /// <param name="grainType">Type of this grain [fully qualified class name]</param>
+        /// <param name="grainReference">Grain reference object for this grain.</param>
+        /// <param name="grainState">State data object to be populated for this grain.</param>
+        /// <returns>Completion promise for the Read operation on the specified grain.</returns>
+        Task ReadStateAsync(string grainType, GrainReference grainReference, IGrainState grainState);
+
+        /// <summary>Write data function for this storage instance.</summary>
+        /// <param name="grainType">Type of this grain [fully qualified class name]</param>
+        /// <param name="grainReference">Grain reference object for this grain.</param>
+        /// <param name="grainState">State data object to be written for this grain.</param>
+        /// <returns>Completion promise for the Write operation on the specified grain.</returns>
+        Task WriteStateAsync(string grainType, GrainReference grainReference, IGrainState grainState);
+
+        /// <summary>Delete / Clear data function for this storage instance.</summary>
+        /// <param name="grainType">Type of this grain [fully qualified class name]</param>
+        /// <param name="grainReference">Grain reference object for this grain.</param>
+        /// <param name="grainState">Copy of last-known state data object for this grain.</param>
+        /// <returns>Completion promise for the Delete operation on the specified grain.</returns>
+        Task ClearStateAsync(string grainType, GrainReference grainReference, IGrainState grainState);
+    }
+
+    /// <summary>
+    /// Interface to be optionally implemented by storage to return richer exception details.
+    /// TODO: Remove this interface.  Move to decorator pattern for monitoring purposes. - jbragg
+    /// </summary>
+    public interface IRestExceptionDecoder
+    {
+        /// <summary>
+        /// Decode details of the exceprion
+        /// </summary>
+        /// <param name="e">Excption to decode</param>
+        /// <param name="httpStatusCode">HTTP status code for the error</param>
+        /// <param name="restStatus">REST status for the error</param>
+        /// <param name="getExtendedErrors">Whether or not to extract REST error code</param>
+        /// <returns></returns>
+        bool DecodeException(Exception e, out HttpStatusCode httpStatusCode, out string restStatus, bool getExtendedErrors = false);
+    }
+
+    /// <summary>
+    /// Exception thrown whenever a grain call is attempted with a bad / missing storage configuration settings for that grain.
+    /// </summary>
+    [Serializable]
+    public class BadStorageConfigException : OrleansException
```

Comment:
maybe name it `BadGrainStorageConfigurationException` ? storage is too broad, that's why you named `IGrainStorage` to be grain specific I think .

Reply:
And avoid using short names like config, instead of configuration, in public class names

Reply:
This was moved from IStorageProvider.  It's a preexisting exception.  If we change the name it could break application code's error handling.  Not 100% against changing the name, but don't see much advantage in it.

Reply:
My mistake, this was added to communicate grain storage issues, I got confused. :/
Good catch, tnx! :)